### ISSUE TITLE
Error: 260002: password is empty

### ISSUE THREAD

**Issue by jifeous** [Snowflake-Labs/terraform-provider-snowflake#2133]
**Provider Version**
0.74.0

**Terraform Version**
Terraform v1.6.1
on windows_amd64
+ provider registry.terraform.io/hashicorp/aws v5.21.0
+ provider registry.terraform.io/snowflake-labs/snowflake v0.74.0

**Describe the bug**
Snowflake provider error after update to 0.74.0 from 0.73.0. Using private key with no password, seems that the provider requires a password and is failing now.

│ Error: 260002: password is empty
│
│   with provider["registry.terraform.io/snowflake-labs/snowflake"],
│   on provider.tf line 29, in provider "snowflake":
│   29: provider "snowflake" {
│

**Expected behavior**
No error

**Comment by simonepm** [comment]
Issue still present in 0.75.0

**Comment by francescomucio** [comment]
Can anyone fix this?
We connect to Snowflake using a private key and this is just breaking our CICD pipeline.

I could try to contribute to this, but I will need some guidance

**Comment by sfc-gh-swinkler** [comment]
Hello, sorry for the issue with provider configuration. We are working on a fix and will have this out ASAP. Thank you for your patience.

**Comment by gcyu** [comment]
Hi @sfc-gh-swinkler thanks for the fix. I am still seeing this issue in v0.75.0. Do you need to recreate the tag or release a new version?

**Comment by joscha** [comment]
I seem to be getting this with `0.96.0` when using:


```
export SNOWFLAKE_USER="snowflake_username"
export SNOWFLAKE_PRIVATE_KEY_PATH="~/.ssh/snowflake_key.p8"
export SNOWFLAKE_PRIVATE_KEY_PASSPHRASE="snowflake_passphrase"
```

and 
```hcl
provider "snowflake" {
  role  = "ORGADMIN"
  alias = "orgadmin"
  account = "..."
}
```
There is also a warning about the deprecation of the `SNOWFLAKE_PRIVATE_KEY_PATH` argument and using `file()` instead. Possibly related?

**Comment by sfc-gh-jmichalak** [comment]
Hi @joscha 👋 
This field is deprecated but should work. Please take a look at solutions in another [issue](https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2169#issuecomment-1816167117), and try using `private_key` instead of `private_key_path`.

**Comment by joscha** [comment]
> try using `private_key` instead of `private_key_path`

I did that, but:
```
2024-09-20T13:43:47.763+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.opentofu.org/snowflake-labs/snowflake/0.96.0/darwin_arm64/terraform-provider-snowflake_v0.96.0 args=[".terraform/providers/registry.opentofu.org/snowflake-labs/snowflake/0.96.0/darwin_arm64/terraform-provider-snowflake_v0.96.0"]
2024-09-20T13:43:47.766+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.opentofu.org/snowflake-labs/snowflake/0.96.0/darwin_arm64/terraform-provider-snowflake_v0.96.0 pid=50709
2024-09-20T13:43:47.766+0200 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.opentofu.org/snowflake-labs/snowflake/0.96.0/darwin_arm64/terraform-provider-snowflake_v0.96.0
2024-09-20T13:43:47.790+0200 [INFO]  provider.terraform-provider-snowflake_v0.96.0: configuring server automatic mTLS: timestamp="2024-09-20T13:43:47.790+0200"
2024-09-20T13:43:47.797+0200 [DEBUG] provider: using plugin: version=6
2024-09-20T13:43:47.797+0200 [DEBUG] provider.terraform-provider-snowflake_v0.96.0: plugin address: address=/var/folders/ym/n59zz1051yb2ls2t67k59mdm0000gn/T/plugin3346491303 network=unix timestamp="2024-09-20T13:43:47.797+0200"
2024-09-20T13:43:47.805+0200 [DEBUG] provider.terraform-provider-snowflake_v0.96.0: 2024/09/20 13:43:47 [DEBUG] No Snowflake config file found, returning empty config: open /Users/joscha/.snowflake/config: no such file or directory
2024-09-20T13:43:47.805+0200 [DEBUG] provider.terraform-provider-snowflake_v0.96.0: 2024/09/20 13:43:47 [DEBUG] Registering snowflake-instrumented driver
2024-09-20T13:43:47.805+0200 [ERROR] provider.terraform-provider-snowflake_v0.96.0: Response contains error diagnostic: tf_rpc=ConfigureProvider @caller=github.com/hashicorp/terraform-plugin-go@v0.22.2/tfprotov6/internal/diag/diagnostics.go:58 tf_proto_version=6.4 diagnostic_summary="260002: password is empty" tf_provider_addr=registry.terraform.io/Snowflake-Labs/snowflake tf_req_id=f66a1a72-f882-dc07-289c-3f098ef593fc @module=sdk.proto diagnostic_detail="" diagnostic_severity=ERROR timestamp="2024-09-20T13:43:47.805+0200"
2024-09-20T13:43:47.805+0200 [ERROR] vertex "provider[\"registry.opentofu.org/snowflake-labs/snowflake\"].orgadmin" error: 260002: password is empty
╷
│ Error: 260002: password is empty
│
│   with provider["registry.opentofu.org/snowflake-labs/snowflake"].orgadmin,
│   on /Users/joscha/dev/infrastructure/data_warehouse/provider_snowflake.tf line 19, in provider "snowflake":
│   19: provider "snowflake" {
```

with:
```hcl
provider "snowflake" {
  role  = "ORGADMIN"
  alias = "orgadmin"
  private_key = file("~/.ssh/my.p8")
  account = "..."
}
```
`SNOWFLAKE_PRIVATE_KEY_PATH` is unset, `SNOWFLAKE_USER` and `SNOWFLAKE_PRIVATE_KEY_PASSPHRASE` still carry values.

There is a
```
[DEBUG] No Snowflake config file found, returning empty config: open /Users/joscha/.snowflake/config: no such file or directory
```
in the above, but I wouldn't think that should stop it from connecting?

**Comment by joscha** [comment]
Even:
```
provider "snowflake" {
  role  = "ORGADMIN"
  alias = "orgadmin"
  private_key = file("~/.ssh/snowflake_planeta.p8")
  private_key_passphrase = var.private_key_passphrase
  account = "..."
}
```
with the passphrase entered directly to make sure it's not an env issue, yields the same:
```
2024-09-20T14:10:32.446+0200 [DEBUG] ReferenceTransformer: "provider[\"registry.opentofu.org/snowflake-labs/snowflake\"].orgadmin" references: [var.private_key_passphrase]

2024-09-20T14:10:32.491+0200 [ERROR] provider.terraform-provider-snowflake_v0.96.0: Response contains error diagnostic: @module=sdk.proto diagnostic_severity=ERROR tf_req_id=ed4f3012-ea92-496f-6d6a-fbbe89585c17 tf_rpc=ConfigureProvider diagnostic_detail="" diagnostic_summary="260002: password is empty" tf_proto_version=6.4 tf_provider_addr=registry.terraform.io/Snowflake-Labs/snowflake @caller=github.com/hashicorp/terraform-plugin-go@v0.22.2/tfprotov6/internal/diag/diagnostics.go:58 timestamp="2024-09-20T14:10:32.491+0200"
2024-09-20T14:10:32.491+0200 [ERROR] vertex "provider[\"registry.opentofu.org/snowflake-labs/snowflake\"].orgadmin" error: 260002: password is empty
╷
│ Error: 260002: password is empty
│
│   with provider["registry.opentofu.org/snowflake-labs/snowflake"].orgadmin,
│   on /Users/joscha/dev/infrastructure/data_warehouse/provider_snowflake.tf line 19, in provider "snowflake":
│   19: provider "snowflake" {
```

**Comment by sfc-gh-jcieslak** [comment]
Hey @joscha 👋 
You can take a look at this: https://quickstarts.snowflake.com/guide/terraforming_snowflake/index.html#0 I believe your config may work now, but mentioned `authenticator = "JWT"` is missing.

**Comment by joscha** [comment]
> but mentioned `authenticator = "JWT"` is missing.

Yes, I saw, but adding that consistently yields:
```
│ Error: open snowflake connection: 390144 (08004): JWT token is invalid. [8e4343f-4172-4b1a-a83b-79691d587503]
│
│   with provider["registry.opentofu.org/snowflake-labs/snowflake"].orgadmin,
│   on /Users/joscha/dev/infrastructure/data_warehouse/provider_snowflake.tf line 19, in provider "snowflake":
│   19: provider "snowflake" {
```

**Comment by sfc-gh-jcieslak** [comment]
Did you generate it with those instructions: https://quickstarts.snowflake.com/guide/terraforming_snowflake/index.html#2? It may be invalid because you used other options to generate it.

**Comment by sfc-gh-jmichalak** [comment]
This error is on the Snowflake side. Could you check if your key format is valid? Go to the [official Snowflake documentation](https://docs.snowflake.com/en/user-guide/key-pair-auth-troubleshooting#list-of-errors) and search by error code (390144 in this case).

**Comment by joscha** [comment]
> Did you generate it with those instructions: https://quickstarts.snowflake.com/guide/terraforming_snowflake/index.html#2? It may be invalid because you used other options to generate it.

Thank you for that. I did not create it after that exactly, I used: https://docs.snowflake.com/en/user-guide/key-pair-auth. I attached the private key (the same way as described) to an existing user. Will give creating a new user a try.

**Comment by joscha** [comment]
> > Did you generate it with those instructions: https://quickstarts.snowflake.com/guide/terraforming_snowflake/index.html#2? It may be invalid because you used other options to generate it.

Just for future reference. If you try to generate an encrypted private key with the options from the Quickstart above, you'll get:
```
│ Error: could not retrieve private key: could not parse encrypted private key with passphrase, only ciphers aes-128-cbc, aes-128-gcm, aes-192-cbc, aes-192-gcm, aes-256-cbc, aes-256-gcm, and des-ede3-cbc are supported err = pkcs8: only PBES2 supported
```
So you can only generate non-encrypted key that way - for encrypted keys and some more background, see [here](https://medium.com/snowflake/how-to-use-an-encrypted-private-key-in-key-pair-authentication-with-terraform-plugin-for-snowflake-ba5ea63278ec).

**Comment by joscha** [comment]
I got it to work eventually. There seems to be some funkiness with the regeneration of `RSA_PUBLIC_KEY_FP` after using:
```
ALTER USER "tf-snow" SET RSA_PUBLIC_KEY='MII [...] ==';
```

I tried to reproduce it but wasn't able to in a consistent manner. It possibly seems to be related to newlines and/or leading or trailing whitespace in the key input key, but sometimes the fingerprint is different / the last value.
